// binary-live-api/lib/liveevents.js
'use strict';

exports.__esModule = true;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var LiveEvents = (function () {
    function LiveEvents() {
        _classCallCheck(this, LiveEvents);

        this.messageHandlers = {
            '*': []
        };
    }

    LiveEvents.prototype.emitSingle = function emitSingle(msgType, msgData) {

        this.messageHandlers[msgType].forEach(function (handler) {
            handler(msgData);
        });
    };

    LiveEvents.prototype.emitWildcard = function emitWildcard(msgData) {
        this.messageHandlers['*'].forEach(function (handler) {
            handler(msgData);
        });
    };

    LiveEvents.prototype.emit = function emit(msgType, msgData) {

        if (!this.messageHandlers[msgType]) return;

        this.emitSingle(msgType, msgData);
        this.emitWildcard(msgData);
    };

    LiveEvents.prototype.on = function on(msgType, callback) {

        if (!this.messageHandlers[msgType]) this.messageHandlers[msgType] = [];

        this.messageHandlers[msgType].push(callback);
    };

    return LiveEvents;
})();

exports['default'] = LiveEvents;
module.exports = exports['default'];
