// binary-live-api/lib/liveapi.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var _LiveEvents = require('./LiveEvents');

var _LiveEvents2 = _interopRequireDefault(_LiveEvents);

var apiUrl = 'wss://ws.binary.com/websockets/v2';

var LiveApi = (function () {
    _createClass(LiveApi, null, [{
        key: 'Status',
        value: {
            Unknown: 'unknown',
            Connected: 'connected'
        },
        enumerable: true
    }]);

    function LiveApi() {
        _classCallCheck(this, LiveApi);

        this.status = LiveApi.Status.Unknown;
        this.bufferedSends = [];
        this.bufferedExecutes = [];

        this.events = new _LiveEvents2['default']();

        this.socket = new WebSocket(apiUrl);
        this.socket.onopen = this.onOpen.bind(this);
        this.socket.onclose = this.onClose.bind(this);
        this.socket.onerror = this.onError.bind(this);
        this.socket.onmessage = this.onMessage.bind(this);
    }

    LiveApi.prototype.isReady = function isReady() {
        return this.socket && this.socket.readyState === 1;
    };

    LiveApi.prototype.sendBufferedSends = function sendBufferedSends() {
        while (this.bufferedSends.length > 0) {
            this.socket.send(JSON.stringify(this.bufferedSends.shift()));
        }
    };

    LiveApi.prototype.executeBufferedExecutes = function executeBufferedExecutes() {
        while (this.bufferedExecutes.length > 0) {
            this.bufferedExecutes.shift()();
        }
    };

    LiveApi.prototype.onOpen = function onOpen() {
        this.sendBufferedSends();
        this.executeBufferedExecutes();
    };

    LiveApi.prototype.onClose = function onClose() {};

    LiveApi.prototype.onError = function onError(error) {
        console.log(error);
    };

    LiveApi.prototype.onMessage = function onMessage(message) {
        var json = JSON.parse(message.data);
        this.events.emit(json.msg_type, {
            type: json.msg_type,
            data: json[json.msg_type],
            echo: json.echo_req
        });
    };

    LiveApi.prototype.send = function send(data) {
        data.uid = Math.random() * 1e17;
        if (this.isReady()) {
            this.socket.send(JSON.stringify(data));
        } else {
            this.bufferedSends.push(data);
        }
    };

    LiveApi.prototype.execute = function execute(func) {
        if (this.isReady()) {
            func();
        } else {
            this.bufferedExecutes.push(func);
        }
    };

    /////

    LiveApi.prototype.getMarketHistory = function getMarketHistory() {
        var marketHistoryOptions = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

        this.send(_extends({
            ticks: symbol
        }, marketHistoryOptions));
    };

    LiveApi.prototype.getActiveSymbolsBrief = function getActiveSymbolsBrief() {
        this.send({
            active_symbols: 'brief'
        });
    };

    LiveApi.prototype.getActiveSymbolsFull = function getActiveSymbolsFull() {
        this.send({
            active_symbols: 'full'
        });
    };

    LiveApi.prototype.getContractsForSymbol = function getContractsForSymbol(symbol) {
        this.send({
            contracts_for: symbol
        });
    };

    LiveApi.prototype.getPayoutCurrencies = function getPayoutCurrencies() {
        this.send({
            payout_currencies: 1
        });
    };

    LiveApi.prototype.getTradingTimes = function getTradingTimes() {
        var date = arguments.length <= 0 || arguments[0] === undefined ? Date.now() : arguments[0];

        var dateStr = date.getFullYear() + '-' + date.getMonth() + '-' + date.getDate();
        this.send({
            trading_times: date.toString(dateStr)
        });
    };

    LiveApi.prototype.ping = function ping() {
        this.send({
            ping: 1
        });
    };

    LiveApi.prototype.getServerTime = function getServerTime() {
        this.send({
            time: 1
        });
    };

    /////

    LiveApi.prototype.subscribeToTick = function subscribeToTick(symbol) {
        this.send({
            ticks: symbol
        });
    };

    LiveApi.prototype.subscribeToTicks = function subscribeToTicks(symbols) {
        symbols.forEach(this.subscribeToTick.bind(this));
    };

    LiveApi.prototype.getLatestPriceForContractProposal = function getLatestPriceForContractProposal(contractProposal) {
        this.send(_extends({
            proposal: 1
        }, contractProposal));
    };

    LiveApi.prototype.unsubscribeFromTick = function unsubscribeFromTick(symbol) {
        this.send({
            forget: symbol
        });
    };

    LiveApi.prototype.unsubscribeFromTicks = function unsubscribeFromTicks(symbols) {
        symbols.forEach(this.unsubscribeFromTick);
    };

    LiveApi.prototype.unsubscribeFromAllTicks = function unsubscribeFromAllTicks() {
        this.send({
            forget_all: "ticks"
        });
    };

    LiveApi.prototype.unsubscribeFromAllProposals = function unsubscribeFromAllProposals() {
        this.send({
            forget_all: "proposal"
        });
    };

    LiveApi.prototype.unsubscribeFromAllPortfolios = function unsubscribeFromAllPortfolios() {
        this.send({
            forget_all: "portfolio"
        });
    };

    LiveApi.prototype.unsubscribeFromAlProposals = function unsubscribeFromAlProposals() {
        this.send({
            forget_all: "proposal_open_contract"
        });
    };

    /////

    LiveApi.prototype.authorize = function authorize(token) {
        this.send({
            authorize: token
        });
    };

    LiveApi.prototype.getBalance = function getBalance() {
        this.send({
            balance: 1
        });
    };

    LiveApi.prototype.getStatement = function getStatement() {
        var statementOptions = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

        this.send(_extends({
            statement: 1
        }, statementOptions));
    };

    LiveApi.prototype.getPortfolio = function getPortfolio(subscribeToUpdates) {
        this.send({
            portfolio: 1,
            spawn: +subscribeToUpdates
        });
    };

    LiveApi.prototype.getPriceForOpenContract = function getPriceForOpenContract(contractId) {
        this.send({
            proposal_open_contract: 1,
            fmd_id: contractId
        });
    };

    LiveApi.prototype.buyContract = function buyContract(contractId, price) {
        this.send({
            buy: contractId,
            price: price
        });
    };

    LiveApi.prototype.sellContract = function sellContract(contractId, price) {
        this.send({
            sell: contractId,
            price: price
        });
    };

    return LiveApi;
})();

exports['default'] = LiveApi;
module.exports = exports['default'];
